### PR TITLE
WT-6904 Architecture guide for transactions

### PR DESCRIPTION
**Pull request by quchenhao**

### CODE REVIEW

**Comment by quchenhao** [comment]
@luke-pearson Here's the first draft. Would be nice if you have any early feedback.

File: src/docs/arch-glossary.dox [hunk 1]
```diff
@@ -24,7 +24,11 @@ tree - it describes terms internal to the storage engine.
 <table>
 <caption id="transaction_terms">Transaction Term Table</caption>
 <tr><th>Term <th>Definition
-<tr><td>Hello<td>The typical next word is World
+<tr><td>transaction<td>A unit of work performed in WiredTiger that is atomic and consistent with the specified isolation levels, and durability levels.
+<tr><td>session<td>The container that managers transactions and performs the transactional operations on behalf of the users in a single thread.
```
Comment (luke-pearson) [inline src/docs/arch-glossary.dox:28]: manages*

File: src/docs/arch-transaction.dox [hunk 2]
```diff
@@ -5,4 +5,114 @@ A caller of WiredTiger uses @ref transactions within the API to start and stop t
 a session (thread of control).
 
 Internally, the current transaction state is represented by the WT_TXN structure.
+
+WiredTiger performs all the operations in transactions. If the user doesn't explicitly declare a transaction, WiredTiger will automatically create a transaction for the user's operation.
+
+@section Lifecycle
+
+A WiredTiger session creates and managers the transactions' whole lifecycle. A session is only able to manage one transaction at any given time and it cannot run another transaction until the current one completes. Since every session are executed on a single thread, all the operations in the transaction are executed sequentially on the same thread as well.
+
```
Comment (luke-pearson) [inline src/docs/arch-transaction.dox:14]: Again, manages.

File: src/docs/arch-transaction.dox [hunk 3]
```diff
@@ -5,4 +5,114 @@ A caller of WiredTiger uses @ref transactions within the API to start and stop t
 a session (thread of control).
 
 Internally, the current transaction state is represented by the WT_TXN structure.
+
+WiredTiger performs all the operations in transactions. If the user doesn't explicitly declare a transaction, WiredTiger will automatically create a transaction for the user's operation.
+
+@section Lifecycle
+
+A WiredTiger session creates and managers the transactions' whole lifecycle. A session is only able to manage one transaction at any given time and it cannot run another transaction until the current one completes. Since every session are executed on a single thread, all the operations in the transaction are executed sequentially on the same thread as well.
```
Comment (luke-pearson) [inline src/docs/arch-transaction.dox:13]: The word whole here isn't necessar.

File: src/docs/arch-transaction.dox [hunk 4]
```diff
@@ -5,4 +5,114 @@ A caller of WiredTiger uses @ref transactions within the API to start and stop t
 a session (thread of control).
 
 Internally, the current transaction state is represented by the WT_TXN structure.
+
+WiredTiger performs all the operations in transactions. If the user doesn't explicitly declare a transaction, WiredTiger will automatically create a transaction for the user's operation.
```
Comment (luke-pearson) [inline src/docs/arch-transaction.dox:9]: Instead of declare I would say something like "if the user doesn't explicitly begin a transaction"

File: src/docs/arch-glossary.dox [hunk 5]
```diff
@@ -24,7 +24,11 @@ tree - it describes terms internal to the storage engine.
...
+<tr><td>atomicity<td>The operations in a single transaction either all occur or nothing occurs after the transaction finishes.
```
Comment (luke-pearson) [inline src/docs/arch-glossary.dox:29]: This definition of atomicity can be improved, I would suggest something like "Atomicity is a guarantee provided within the context of a transaction that all operations made within that transaction either succeed or fail."

File: src/docs/arch-glossary.dox [hunk 6]
```diff
@@ -24,7 +24,11 @@ tree - it describes terms internal to the storage engine.
...
+<tr><td>isolation<td>Isolation determines whether a transaction should see the operations done by other concurrent transactions.
```
Comment (luke-pearson) [inline src/docs/arch-glossary.dox:30]: I think this definition can be improved too: "Isolation determines which versions of data are visible to a running transaction, and whether it can see changes made by concurrently running transactions."

File: src/docs/arch-transaction.dox (hunk 4 above)
Comment (luke-pearson) [reply src/docs/arch-transaction.dox:9]: Another comment on this line, I think we should be specific about which operations WiredTiger will internally begin a transaction for, do we start one for a single cursor read? If so which isolation level will it run at. These things I think will be helpful to have specified.

File: src/docs/arch-transaction.dox (hunk 3 above)
Comment (luke-pearson) [reply src/docs/arch-transaction.dox:13]: Another comment for this line, I think the wording is a bit odd, I would phrase this as "One one transaction can be run at a time per session, and that transaction must complete before another transaction can be started." or something. Additionally "Since every session are executed on a single thread" seems grammatically weird. "Since every session is singly-threaded all the operations in the transaction are executed on the same thread".

File: src/docs/arch-transaction.dox [hunk 7]
```diff
@@ -5,4 +5,114 @@ A caller of WiredTiger uses @ref transactions within the API to start and stop t
...
+    :rollback;
+split again
+    :commit;
+end split
+
+Stop
+@enduml
+@plantuml_end
+
+A transaction starts when the user explicitly declares
```
Comment (luke-pearson) [inline src/docs/arch-transaction.dox:50]: Whitespace on this line can be fixed.

Comment (luke-pearson) [reply src/docs/arch-transaction.dox:50]: Probably worth being specific here and saying "A transaction starts in two scenarios, when the user calls begin via WT_SESSION::begin_transaction or internally when the user preforms and operation. Internally they are only started if they are not already within the context of a running transaction"

Comment (luke-pearson) [reply src/docs/arch-transaction.dox:50]: "If declared explicitly the transaction will be active until..."

Comment (luke-pearson) [reply src/docs/arch-transaction.dox:50]: Given that we talk about the transaction being started internally for operations a few times, we need to specify which operations either in the previous mention or here in my opinion.

File: src/docs/arch-transaction.dox (hunk 4 above)
Comment (quchenhao) [reply src/docs/arch-transaction.dox:9]: Some of them are mentioned later in the isolation section.

**Comment by luke-pearson** [comment]
It appears that the architecture pages are intended to be word wrapped at 100 characters, see the arch-eviction.dox page for an example. Could you go through and wrap this page?

**Comment by luke-pearson** [comment]
I'm don't think this page is where we should talk about timestamps, and in particular out of order timestamps, it was my understanding that the intent of the pages was to give the developers some insight into the implementation and concepts related to the page. Timestamps feel like a separate concept to me.

**Comment by quchenhao** [comment]
> I'm don't think this page is where we should talk about timestamps, and in particular out of order timestamps, it was my understanding that the intent of the pages was to give the developers some insight into the implementation and concepts related to the page. Timestamps feel like a separate concept to me.

Without talking about timestamps, it is very hard to explain transaction visibility. At least I need to mention it somehow.

**Comment by luke-pearson** [comment]
> Without talking about timestamps, it is very hard to explain transaction visibility. At least I need to mention it somehow.

Well I think it should be mentioned lightly and not have any reference to out of order timestamps.

**Comment by quchenhao** [comment]
> > Without talking about timestamps, it is very hard to explain transaction visibility. At least I need to mention it somehow.
> 
> Well I think it should be mentioned lightly and not have any reference to out of order timestamps.

There is no arch page for timestamp so I think we should mention it here.

**Comment by luke-pearson** [comment]
@ddanderson could you let us know whether you think timestamps would be included in a transaction page and if so to what level of detail?

**Comment by ddanderson** [comment]
@quchenhao , @luke-pearson , Given that we don't have a "home" in the architecture guide for timestamps, it's totally appropriate to talk about it here.  If (later) we find that we need to expand this timestamp section, we can consider giving it its own page.  I've read through the changes and it feels like the right level of detail.  Remember also that there is API level description of timestamp here: https://source.wiredtiger.com/develop/transactions.html#transaction_timestamps, perhaps it's appropriate to reference that as well (via `@ref transaction_timestamps`).  You never know what level of knowledge the reader is starting with.

File: src/docs/arch-transaction.dox [hunk 8]
```diff
@@ -1,8 +1,202 @@
...
+visible update is found. The in-memory updates in WiredTiger are organized as a singly linked
+list with the newest update at the head, called the update chain. If no value is visible on the
+update chain, it checks the version on the disk image, which is the version that was chosen to
+be written to disk in the last reconciliation. If it is still invisible, WiredTiger will search
+the history store to check if there is a version visible to the reader there.
+
+The repeated read guarantee under snapshot isolation may break in one case if the timestamps
+committed to the updates are out of order, e.g,
+
+U@20 -> U@30 -> U@15.
```
Comment (ddanderson) [inline src/docs/arch-transaction.dox:163]: Tip: To put this paragraph in monospace font, use three backticks - an illustration below with backtick (`) replaced by ~ (since github also treats backtick specially):
```
~~~
U@20 ....
~~~
```

File: src/docs/arch-transaction.dox [hunk 9]
```diff
@@ -1,8 +1,202 @@
...
+
+In the above example, reading with timestamp 15 doesn't guarantee to return the third update.
+In some cases, users may read the second update U@30 if it is moved to the history store.
```
Comment (ddanderson) [inline src/docs/arch-transaction.dox:166]: `U@30` -> `\c U@30`. (The `\c` puts the next item into monospace font).

File: src/docs/arch-transaction.dox [hunk 10]
```diff
@@ -1,8 +1,202 @@
...
+
+@subsection Durability
+
+WiredTiger transactions support commit level durability and checkpoint level durability. An
+operation is commit level durable if logging is enabled on the table (@ref arch-logging). After
+it has been successfully commit, the operation is guaranteed to survive restart. An operation
```
Comment (ddanderson) [inline src/docs/arch-transaction.dox:172]: `commit` -> `committed`

File: src/docs/arch-transaction.dox [hunk 11]
```diff
@@ -1,8 +1,202 @@
...
+will only survive across restart under checkpoint durability if it is included in the last
+successful checkpoint.
+
+@section Prepared Transaction
```
Comment (ddanderson) [inline src/docs/arch-transaction.dox:176]: I'd make this `Transactions`, it reads better.

File: src/docs/arch-transaction.dox [hunk 12]
```diff
@@ -1,8 +1,202 @@
...
+
+WiredTiger introduces prepared transaction to meet the needs of implementing distributed
```
Comment (ddanderson) [inline src/docs/arch-transaction.dox:178]: `transaction` -> `transactions`

File: src/docs/arch-transaction.dox [hunk 13]
```diff
@@ -1,8 +1,202 @@
...
+transactions through two-phase commit. Prepared transactions only work under snapshot isolation.
+
+Instead of just having the beginning, operating, and rollback or commit phase, it has a
+prepared phase before the rollback or commit phase. After prepare is called, WiredTiger
+releases the transaction's snapshot and prohibit any more read or write operations on the
```
Comment (ddanderson) [inline src/docs/arch-transaction.dox:183]: `prohibit` -> `prohibits`

**Review by ddanderson — APPROVED** [review]
I really liked reading this, great content!  I'm approving with a number of "proofreader" corrections.

File: src/docs/arch-transaction.dox [hunk 14]
```diff
@@ -1,8 +1,202 @@
...
-Transactions provide a powerful abstraction for multiple threads to operate on data concurrently.
-A caller of WiredTiger uses @ref transactions within the API to start and stop transactions within
-a session (thread of control).
+Transactions provide a powerful abstraction for multiple threads to operate on data
+ concurrently. A caller of WiredTiger uses @ref transactions within the API to start and stop
+ transactions within a session (thread of control).
 
 Internally, the current transaction state is represented by the WT_TXN structure.
+
+WiredTiger performs all the operations in transactions. If the user doesn't explicitly begin a
```
Comment (luke-pearson) [inline src/docs/arch-transaction.dox:9]: This reads weirdly, "WiredTiger performs all operations within a transaction". That isn't true for operations like schema drop, create etc. How about "WiredTiger peforms most operations within a transaction" or similar?

File: src/docs/arch-transaction.dox [hunk 15]
```diff
@@ -1,8 +1,202 @@
...
+checkpoint timestamp in the transaction will be discarded.
+
+There is another case that atomicity may be violated if a transaction operates both on tables
+with logging enabled and disabled after restart. The operations on the tables with logging
+enabled will survive the restart, while the operations on the non-logged tables may be lost if
+it is not included in the latest checkpoint.
+
+@subsection Isolation
+
+Isolation is one of the important feature of database to decide whether one transaction can
```
Comment (luke-pearson) [inline src/docs/arch-transaction.dox:95]: feature -> features

File: src/docs/arch-transaction.dox [hunk 16]
```diff
@@ -1,8 +1,202 @@
...
+before calling another search. Read uncommitted transactions don't have a snapshot.
+
+If the transaction has a snapshot, each read will check whether the update's transaction id is
+in its snapshot. The updates with transaction ids in the snapshot or larger than the largest
+transaction id in the snapshot are not visible to the reading transaction.
+
+When operating in read committed or read uncommitted isolation levels, it is possible to read
+different values of the same key, seeing records not seen before, or finding records disappear
+in the same transaction. This is called a phantom read. Under snapshot isolation, WiredTiger guarantees
+repeated reads returning the same result except one case using timestamps.
```
Comment (luke-pearson) [inline src/docs/arch-transaction.dox:126]: except in one scenario using timestamps.

File: src/docs/arch-transaction.dox [hunk 17]
```diff
@@ -1,8 +1,202 @@
...
+prepared phase before the rollback or commit phase. After prepare is called, WiredTiger
+releases the transaction's snapshot and prohibits any more read or write operations on the
+transaction.
+
+By introducing the prepared stage, a two-phase distributed transaction algorithm can rely on
+the prepared state to reach consensus among all the nodes for committing.
+
+Along with the prepared phase, WiredTiger introduces the prepared timestamp and durable
+timestamp. They are to prevent the slow prepared transactions blocking the movement of the
+global stable timestamp, which may cause excessive amount of data being pinned in memory. The
```
Comment (luke-pearson) [inline src/docs/arch-transaction.dox:191]: which may cause excessive amounts of data to be pinned in memory.

File: src/docs/arch-transaction.dox [hunk 18]
```diff
@@ -1,8 +1,202 @@
...
+
+Instead of just having the beginning, operating, and rollback or commit phase, it has a
+prepared phase before the rollback or commit phase. After prepare is called, WiredTiger
+releases the transaction's snapshot and prohibits any more read or write operations on the
+transaction.
+
+By introducing the prepared stage, a two-phase distributed transaction algorithm can rely on
+the prepared state to reach consensus among all the nodes for committing.
+
+Along with the prepared phase, WiredTiger introduces the prepared timestamp and durable
```
Comment (luke-pearson) [inline src/docs/arch-transaction.dox:189]: Seeing as we're talking about the timestamps here can we define the guarantee wiredtiger provides with regards to the durable timestamp?

File: src/docs/arch-transaction.dox [hunk 19]
```diff
@@ -1,8 +1,202 @@
...
+global stable timestamp, which may cause excessive amount of data being pinned in memory. The
+stable timestamp is allowed to move beyond the prepared timestamp and at the commit time, the
+prepared transaction can then be committed after the current stable timestamp with a larger
+durable timestamp.
+
+The visibility of the prepared transaction is also special when in the prepared state. Since in
+the prepared state, the transaction has released its snapshot, it should be visible to the
+transactions starting after that based on the normal visibility rule. However, the prepared
+transaction has not been committed and cannot be visible yet. In this situation, WiredTiger
+either ignores the prepared update and tries to read the older updates if it is configured to
```
Comment (luke-pearson) [inline src/docs/arch-transaction.dox:200]: I would reword this section to be:
In this situation WiredTiger will return a WT_PREPARE_CONFLICT to indicate to the caller to retry later, or if configured WiredTiger will ignore the prepared update and read older updates.

WiredTiger either ignores the prepared update and tries to read older updates if it is configured to

File: src/docs/arch-transaction.dox [hunk 20]
```diff
@@ -92,7 +91,7 @@ it is not included in the latest checkpoint.
 
 @subsection Isolation
 
-Isolation is one of the important feature of database to decide whether one transaction can
+Isolation is one of the important features of database to decide whether one transaction can
```
Comment (luke-pearson) [inline src/docs/arch-transaction.dox:94]: Sorry, I know i got you to adjust this but it should be "features of a database, which is used to determine whether on transaction..."

**Review by luke-pearson — APPROVED** [review]
LGTM, left one final comment.
